File: python/promptsafe/logger.py
"""
PromptSafe — Attack Logger (Python)
Logs detected prompt injection attacks to a local SQLite database.
Zero external dependencies — uses stdlib sqlite3.
"""

import sqlite3
import json
import os
import time
from typing import Any, Dict, List, Optional


_DEFAULT_DB = os.path.join(
    os.path.expanduser("~"),
    ".promptsafe",
    "attacks.db",
)


class AttackLogger:
    """Thread-safe local SQLite logger for prompt injection events."""

    def __init__(self, db_path: Optional[str] = None):
        self.db_path = db_path or _DEFAULT_DB
        os.makedirs(os.path.dirname(self.db_path), exist_ok=True)
        self._init_db()

    # ── Internals ────────────────────────────────────────────────────────────

    def _connect(self) -> sqlite3.Connection:
        conn = sqlite3.connect(self.db_path, timeout=10)
        conn.row_factory = sqlite3.Row
        return conn

    def _init_db(self) -> None:
        with self._connect() as conn:
            conn.execute("""
                CREATE TABLE IF NOT EXISTS attacks (
                    id          INTEGER PRIMARY KEY AUTOINCREMENT,
                    timestamp   TEXT    NOT NULL,
                    input       TEXT    NOT NULL,
                    score       INTEGER NOT NULL,
                    risk_level  TEXT    NOT NULL,
                    reasons     TEXT,
                    is_blocked  INTEGER DEFAULT 1
                )
            """)
            conn.execute(
                "CREATE INDEX IF NOT EXISTS idx_risk ON attacks(risk_level)"
            )
            conn.commit()

    # ── Public API ───────────────────────────────────────────────────────────

    def log(self, event: Dict[str, Any]) -> None:
        """
        Persist an attack event.

        Args:
            event: dict with keys: input, score, riskLevel, reasons, isBlocked
        """
        ts = time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
        reasons_json = json.dumps(event.get("reasons", []))
        with self._connect() as conn:
            conn.execute(
                """
                INSERT INTO attacks (timestamp, input, score, risk_level, reasons, is_blocked)
                VALUES (?, ?, ?, ?, ?, ?)
                """,
                (
                    ts,
                    (event.get("input", "") or "")[:500],
                    int(event.get("score", 0)),
                    event.get("riskLevel", "unknown"),
                    reasons_json,
                    1 if event.get("isBlocked", True) else 0,
                ),
            )
            conn.commit()

    def get_recent(
        self,
        limit: int = 50,
        risk_level: Optional[str] = None,
    ) -> List[Dict[str, Any]]:
        """Return recent log entries, newest first."""
        query = "SELECT * FROM attacks"
        params: list = []
        if risk_level:
            query += " WHERE risk_level = ?"
            params.append(risk_level)
        query += " ORDER BY id DESC LIMIT ?"
        params.append(limit)

        with self._connect() as conn:
            rows = conn.execute(query, params).fetchall()

        result = []
        for row in rows:
            result.append({
                "id": row["id"],
                "timestamp": row["timestamp"],
                "input": row["input"],
                "score": row["score"],
                "riskLevel": row["risk_level"],
                "reasons": json.loads(row["reasons"] or "[]"),
                "isBlocked": bool(row["is_blocked"]),
            })
        return result

    def get_stats(self) -> Dict[str, Any]:
        """Return aggregate statistics."""
        with self._connect() as conn:
            total = conn.execute("SELECT COUNT(*) FROM attacks").fetchone()[0]
            blocked = conn.execute(
                "SELECT COUNT(*) FROM attacks WHERE is_blocked = 1"
            ).fetchone()[0]
            by_risk = conn.execute(
                "SELECT risk_level, COUNT(*) as cnt FROM attacks GROUP BY risk_level"
            ).fetchall()

        return {
            "totalScanned": total,
            "totalBlocked": blocked,
            "totalSafe": total - blocked,
            "byRiskLevel": {row["risk_level"]: row["cnt"] for row in by_risk},
            "dbFile": self.db_path,
        }

    def clear(self) -> int:
        """Delete all log entries. Returns number of rows deleted."""
        with self._connect() as conn:
            cur = conn.execute("DELETE FROM attacks")
            conn.commit()
            return cur.rowcount

File: npm/src/logger.js
'use strict';

/**
 * PromptSafe — Attack Logger (Node.js)
 * Logs detected prompt injection attacks to a local JSON file.
 * Zero external dependencies.
 */

const fs = require('fs');
const path = require('path');
const os = require('os');

const DEFAULT_LOG_FILE = path.join(os.homedir(), '.promptsafe', 'attacks.json');

class AttackLogger {
  /**
   * @param {string} [logFile] - Path to the JSON log file.
   */
  constructor(logFile) {
    this.logFile = logFile || DEFAULT_LOG_FILE;
    this._ensureDir();
  }

  // ── Internals ──────────────────────────────────────────────────────────────

  _ensureDir() {
    const dir = path.dirname(this.logFile);
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }
  }

  _readAll() {
    if (!fs.existsSync(this.logFile)) return [];
    try {
      return JSON.parse(fs.readFileSync(this.logFile, 'utf8'));
    } catch {
      return [];
    }
  }

  _writeAll(entries) {
    fs.writeFileSync(this.logFile, JSON.stringify(entries, null, 2), 'utf8');
  }

  // ── Public API ─────────────────────────────────────────────────────────────

  /**
   * Append an attack event to the log.
   * @param {{ input, score, riskLevel, reasons, isBlocked }} event
   */
  log(event) {
    const entries = this._readAll();
    entries.unshift({
      id: Date.now(),
      timestamp: new Date().toISOString(),
      input: (event.input || '').slice(0, 500),
      score: event.score || 0,
      riskLevel: event.riskLevel || 'unknown',
      reasons: event.reasons || [],
      isBlocked: event.isBlocked !== false,
    });
    // Keep last 1000 entries
    this._writeAll(entries.slice(0, 1000));
  }

  /**
   * Get recent log entries, newest first.
   * @param {number} [limit=50]
   * @param {string|null} [riskLevel]
   * @returns {Object[]}
   */
  getRecent(limit = 50, riskLevel = null) {
    let entries = this._readAll();
    if (riskLevel) {
      entries = entries.filter(e => e.riskLevel === riskLevel);
    }
    return entries.slice(0, limit);
  }

  /**
   * Get aggregate statistics.
   * @returns {{ totalScanned, totalBlocked, totalSafe, byRiskLevel, logFile }}
   */
  getStats() {
    const entries = this._readAll();
    const byRisk = {};
    let blocked = 0;
    for (const e of entries) {
      if (e.isBlocked) blocked++;
      byRisk[e.riskLevel] = (byRisk[e.riskLevel] || 0) + 1;
    }
    return {
      totalScanned: entries.length,
      totalBlocked: blocked,
      totalSafe: entries.length - blocked,
      byRiskLevel: byRisk,
      logFile: this.logFile,
    };
  }

  /**
   * Delete all log entries.
   * @returns {number} Number of entries deleted.
   */
  clear() {
    const count = this._readAll().length;
    this._writeAll([]);
    return count;
  }
}

module.exports = AttackLogger;
